server/src/main/java/io/deephaven/server/partitionedtable/PartitionedTableServiceGrpcImpl.java

```diff
@@ -199,4 +169,84 @@ public void getTable(
                     });
         }
     }
+
+    @TestUseOnly
+    Table lockAndGetConstituents(@NotNull final GetTableRequest request, final Table keyTable,
+            final PartitionedTable partitionedTable) {
+        final boolean requiresLock = keyTable.isRefreshing() || partitionedTable.table().isRefreshing();
+
+        if (requiresLock) {
+            // validate the update graphs are the same
+            final UpdateGraph keyUpdateGraph = keyTable.getUpdateGraph();
+            final UpdateGraph ptUpdateGraph = partitionedTable.table().getUpdateGraph();
+            if (keyUpdateGraph != null && ptUpdateGraph != null && ptUpdateGraph != keyUpdateGraph) {
+                throw Exceptions.statusRuntimeException(Code.INVALID_ARGUMENT,
+                        "Provided key table UpdateGraph is inconsistent with PartitionedTable UpdateGraph");
+            }
+            if (keyUpdateGraph != null) {
+                return keyUpdateGraph.sharedLock()
+                        .computeLocked(() -> getConstituents(request, keyTable, partitionedTable));
+            } else {
+                return ptUpdateGraph.sharedLock()
+                        .computeLocked(() -> getConstituents(request, keyTable, partitionedTable));
+            }
+        } else {
+            return getConstituents(request, keyTable, partitionedTable);
+        }
+    }
+
+    @TestUseOnly
+    Table getConstituents(@NotNull GetTableRequest request, Table keyTable, PartitionedTable partitionedTable) {
+        final boolean uniqueStaticResult;
+
+        switch (request.getUniqueBehavior()) {
```

Comment:
I'm a bit sad, our getTable impl is not calling `io.deephaven.server.grpc.GrpcErrorHelper#checkHasNoUnknownFieldsRecursive`. This means that if a client sets `PERMIT_MULTIPLE_KEYS = 2` and calls an older server, things will silently be incorrect instead of a proper error message.

Reply:
Added to the methods in this service.

Reply:
Thanks. This is one of those cases where I'm glad you added it here, but if we think it's important, we may want to backport just the check to old releases we care about.